go/types/walk_test.go

```diff
@@ -249,3 +250,118 @@ func (suite *WalkTestSuite) SetupTest() {
 	suite.deadValue = Number(0xDEADBEEF)
 	suite.mustSkip = NewList(suite.deadValue)
 }
+
+type WalkDifferentStructsTestSuite struct {
+	suite.Suite
+	vs *ValueStore
+	ts *chunks.TestStore
+}
+
+func TestWalkDifferentStructsTestSuite(t *testing.T) {
+	suite.Run(t, &WalkDifferentStructsTestSuite{})
+}
+
+func (suite *WalkDifferentStructsTestSuite) SetupTest() {
+	suite.ts = chunks.NewTestStore()
+	suite.vs = NewValueStore(NewBatchStoreAdaptor(suite.ts))
+}
+
+func (suite *WalkDifferentStructsTestSuite) AssertDiffs(last, current Value, expectAdded, expectRemoved []Value) {
+	equalIgnoreOrder := func(expect []Value, actual map[hash.Hash]Struct) {
+		hashes := hash.HashSet{}
+		for _, v := range expect {
+			hashes.Insert(v.Hash())
+		}
+
+		suite.Equal(len(hashes), len(actual))
+		for h, _ := range actual {
+			_, ok := hashes[h]
+			suite.True(ok)
+		}
+	}
+
+	added, removed := WalkDifferentStructs(last, current, suite.vs)
+	equalIgnoreOrder(expectAdded, added)
+	equalIgnoreOrder(expectRemoved, removed)
+}
+
+func (suite *WalkDifferentStructsTestSuite) TestWalkStructsBasic() {
+	// Nothing plus nothing is nothing
+	suite.AssertDiffs(nil, nil, []Value{}, []Value{})
+
+	s1 := NewStruct("s", StructData{
+		"n": Number(1),
+	})
+
+	// Top level struct.
+	suite.AssertDiffs(nil, s1, []Value{s1}, []Value{})
+
+	// Same Struct
+	suite.AssertDiffs(s1, s1, []Value{}, []Value{})
+
+	// Removed.
+	suite.AssertDiffs(s1, nil, []Value{}, []Value{s1})
+
+	// Collections of structs
+	s2 := NewStruct("s", StructData{
+		"n": Number(2),
+	})
+	s3 := NewStruct("s", StructData{
+		"n": Number(3),
+	})
+	s4 := NewStruct("s", StructData{
+		"n": Number(4),
+	})
+
+	l1 := NewList(s2, s4)
+	l2 := NewList(s1, s2, s3)
+
+	suite.AssertDiffs(l1, l2, []Value{s1, s3}, []Value{s4})
+	suite.AssertDiffs(l1, l1, []Value{}, []Value{})
+	suite.AssertDiffs(l2, l1, []Value{s4}, []Value{s1, s3})
+
+	// Big, uncommitted collections of structs
+	bl := NewList()
+	i := 0
+	for NewRef(bl).Height() == 1 {
+		bl = bl.Append(Number(i))
+		i++
+	}
+
+	l1 = bl.Append(s2, s4)
+	l2 = bl.Append(s1, s2, s3)
+
+	suite.AssertDiffs(l1, l2, []Value{s1, s3}, []Value{s4})
+	suite.AssertDiffs(l1, l1, []Value{}, []Value{})
+	suite.AssertDiffs(l2, l1, []Value{s4}, []Value{s1, s3})
+
+	// Big, committed collections of structs
+	l1 = suite.vs.ReadValue(suite.vs.WriteValue(l1).TargetHash()).(List)
```

Comment:
I think you mean "flushed" rather than "committed"? Given the grandchild strategy, I think that we won't even have flushed any parts of `l1` or `l2`. They're only of height 2, right?

If you actually are testing the case where chunks of these prolly trees have gotten written out to the batch store layer, then you'll want to do something like 
```golang
l1Hash := suite.vs.WriteValue(l1).TargetHash()
suite.vs.Flush(l1Hash)
l1 = suite.vs.ReadValue(l1Hash).(List)
```